src/simulator/particle/LinkedCellContainer.cpp

```diff
@@ -0,0 +1,372 @@
+  auto particles = _particle_container;
+  clear();
+  for (auto &p : particles) {
+    insert(*p);
+  }
+  logger.debug("New low left: " + std::to_string(left_corner_coordinates[0]) +
+               " " + std::to_string(left_corner_coordinates[1]) + " " +
+               std::to_string(left_corner_coordinates[2]));
+}
+
+void LinkedCellContainer::handleBoundaryConditions(Particle &p) {
+  auto position = p.getX();
+  auto velocity = p.getV();
+
+  // left boundary
+  if (position[0] < left_corner_coordinates[0]) {
+    if (boundary_conditions_.left == BoundaryCondition::Reflecting) {
+      position[0] = 2 * left_corner_coordinates[0] - position[0];
+      velocity[0] = -velocity[0];
+    } else if (boundary_conditions_.left == BoundaryCondition::Outflow) {
+      p.markForRemoval();
+    }
+  }
+
+  // Right boundary
+  if (position[0] > left_corner_coordinates[0] + domain_size_[0]) {
+    if (boundary_conditions_.right == BoundaryCondition::Reflecting) {
+      position[0] =
+          2 * (left_corner_coordinates[0] + domain_size_[0]) - position[0];
+      velocity[0] = -velocity[0];
+    } else if (boundary_conditions_.right == BoundaryCondition::Outflow) {
+      p.markForRemoval();
+    }
+  }
+
+  // Bottom boundary
+  if (position[1] < left_corner_coordinates[1]) {
+    if (boundary_conditions_.bottom == BoundaryCondition::Reflecting) {
+      position[1] = 2 * left_corner_coordinates[1] - position[1];
+      velocity[1] = -velocity[1];
+    } else if (boundary_conditions_.bottom == BoundaryCondition::Outflow) {
+      p.markForRemoval();
+    }
+  }
+
+  // Top boundary
+  if (position[1] > left_corner_coordinates[1] + domain_size_[1]) {
+    if (boundary_conditions_.top == BoundaryCondition::Reflecting) {
+      position[1] =
+          2 * (left_corner_coordinates[1] + domain_size_[1]) - position[1];
+      velocity[1] = -velocity[1];
+    } else if (boundary_conditions_.top == BoundaryCondition::Outflow) {
+      p.markForRemoval();
+    }
+  }
+
+  if (domain_size_.size() > 2) {
+    // Front boundary
+    if (position[2] < left_corner_coordinates[2]) {
+      if (boundary_conditions_.front == BoundaryCondition::Reflecting) {
+        position[2] = 2 * left_corner_coordinates[2] - position[2];
+        velocity[2] = -velocity[2];
+      } else if (boundary_conditions_.front == BoundaryCondition::Outflow) {
+        p.markForRemoval();
+      }
+    }
+    // Back boundary
+    if (position[2] > left_corner_coordinates[2] + domain_size_[2]) {
+      if (boundary_conditions_.back == BoundaryCondition::Reflecting) {
+        position[2] =
+            2 * (left_corner_coordinates[2] + domain_size_[2]) - position[2];
+        velocity[2] = -velocity[2];
+      } else if (boundary_conditions_.back == BoundaryCondition::Outflow) {
+        p.markForRemoval();
+      }
+    }
+  }
+
+  p.updateX(position[0], position[1], position[2]);
+  p.updateV(velocity[0], velocity[1], velocity[2]);
```

Comment:
while it is fine, I feel like this can be optimized (maybe consider as an optimization in WS4) by minimizing the number of if-else loops as this will be done at every step. You can also leave it as it is if there are other things you would like to do first, it is nothing urgent.